uchar.c

```diff
@@ -178,93 +179,24 @@ size_t u_strlen_safe(const char *str)
 
 int u_char_width(uchar u)
 {
-	if (unlikely(!using_utf8))
-		goto narrow;
+	int w;
 
 	if (unlikely(u < 0x20))
 		goto control;
 
-	/* Combining Diacritical Marks */
-	if (u >= 0x300U && u <= 0x36fU)
-		goto zero;
-
-	if (u < 0x1100U)
-		goto narrow;
-
-	/* Hangul Jamo init. consonants */
-	if (u <= 0x115fU)
-		goto wide;
-
-	/* Zero-width characters */
-	if (u == 0x200bU || u == 0x200cU || u == 0x200dU)
-		goto zero;
-
-	/* angle brackets */
-	if (u == 0x2329U || u == 0x232aU)
-		goto wide;
-
-	if (u < 0x2e80U)
-		goto narrow;
-	/* CJK ... Yi */
-	if (u < 0x302aU)
-		goto wide;
-	if (u <= 0x302fU)
-		goto narrow;
-	if (u == 0x303fU)
-		goto narrow;
-	if (u == 0x3099U)
-		goto narrow;
-	if (u == 0x309aU)
-		goto narrow;
-	/* CJK ... Yi */
-	if (u <= 0xa4cfU)
-		goto wide;
-
-	/* Hangul Syllables */
-	if (u >= 0xac00U && u <= 0xd7a3U)
-		goto wide;
-
-	/* CJK Compatibility Ideographs */
-	if (u >= 0xf900U && u <= 0xfaffU)
-		goto wide;
-
-	/* CJK Compatibility Forms */
-	if (u >= 0xfe30U && u <= 0xfe6fU)
-		goto wide;
-
-	/* Fullwidth Forms */
-	if (u >= 0xff00U && u <= 0xff60U)
-		goto wide;
-
-	/* Halfwidth Forms */
-	if (u >= 0xff61U && u <= 0xffdfU)
-		goto narrow;
-
-	/* Fullwidth Forms */
-	if (u >= 0xffe0U && u <= 0xffe6U)
-		goto wide;
-
-	/* Halfwidth Forms */
-	if (u >= 0xffe8U && u <= 0xffeeU)
-		goto narrow;
-
-	/* CJK extra stuff */
-	if (u >= 0x20000U && u <= 0x2fffdU)
-		goto wide;
-
-	/* ? */
-	if (u >= 0x30000U && u <= 0x3fffdU)
-		goto wide;
+	if (unlikely(!using_utf8))
+		return 1;
 
 	/* invalid bytes in unicode stream are rendered "<xx>" */
 	if (u & U_INVALID_MASK)
 		goto invalid;
-zero:
-	return 0;
-narrow:
-	return 1;
-wide:
-	return 2;
+
+	w = wcwidth_uchar(u);
+	if (w >= 0)
+		return w;
+	else
+		return 2;
```

Comment:
Assuming it's for the likes of "￾" would 1 not make more sense? (Assuming we ever get here.)

Reply:
That would be my assumption, but my reading of the original source for this function lead me to believe it returned 2 for unknown characters, for reasons that aren't really clear to me:

```
	/* ? */
	if (u >= 0x30000U && u <= 0x3fffdU)
		goto wide;
```

Maybe the intent was that overallocating space would be preferred to underallocating? I don't have a strong opinion either way, as I believe this should only be reached for private use characters, which is probably better set to a default width of 1, since they are unlikely to show up in song metadata, and if the user wants to use them in their status bar, they can manually pad it if need be.

Reply:
It's worth noting that it's not incredibly uncommon for `wcwidth` to be overriden with `LD_PRELOAD` to handle icon widths as well, for example https://github.com/powerman/wcwidth-icons.

I will change the default width to 1 unless there was a specific reason for the old behavior.

Reply:
I would have taken `/* ? */` to mean "I don't know what to call this range but it seems to be filled with wide characters".

At least [U+FFFE](https://util.unicode.org/UnicodeJsps/character.jsp?a=fffe&B1=Show) does hit this case for me and displays as only 1 column wide, so I'd go with 1.

> It's worth noting that it's not incredibly uncommon for wcwidth to be overriden with LD_PRELOAD to handle icon widths as well

Not something I would have thought of, but I suppose it shouldn't matter to us? If anything this change might let it work properly.

Reply:
This should be resolved by 9a9d878.